packages/core/core.ts

```diff
@@ -41,94 +48,109 @@ declare global {
 	var HookManager: hooking.HookManager;
 }
 
-export function registerGlobals() {
-	globalThis.Fuzzer = fuzzer.fuzzer;
-	//TODO make sure that all sanitizers are registered at this point
-	globalThis.HookManager = hooking.hookManager;
-}
-
 export async function initFuzzing(options: Options) {
 	registerGlobals();
-
 	await Promise.all(options.customHooks.map(importModule));
-
 	if (!options.dryRun) {
 		registerInstrumentor(options.includes, options.excludes);
 	}
 }
 
-async function loadFuzzFunction(options: Options): Promise<fuzzer.FuzzTarget> {
-	const fuzzTarget = await importModule(options.fuzzTarget);
-	if (!fuzzTarget) {
-		throw new Error(
-			`${options.fuzzTarget} could not be imported successfully"`
-		);
-	}
-	const fuzzFn: fuzzer.FuzzTarget = fuzzTarget[options.fuzzEntryPoint];
-	if (typeof fuzzFn !== "function") {
-		throw new Error(
-			`${options.fuzzTarget} does not export function "${options.fuzzEntryPoint}"`
-		);
-	}
-	return fuzzFn;
+export function registerGlobals() {
+	globalThis.Fuzzer = fuzzer.fuzzer;
+	//TODO make sure that all sanitizers are registered at this point
+	globalThis.HookManager = hooking.hookManager;
 }
 
-export async function startFuzzing(options: Options) {
-	await initFuzzing(options);
-	const fuzzFn: fuzzer.FuzzTarget = await loadFuzzFunction(options);
-	startFuzzingNoInit(
-		fuzzFn,
-		addFuzzerOptionsForDryRun(options.fuzzerOptions, options.dryRun)
+export function startFuzzing(options: Options) {
+	const fuzzing = options.sync
+		? startFuzzingSync(options)
+		: startFuzzingAsync(options);
+	fuzzing.then(
+		() => {
+			stopFuzzing(undefined, options.expectedErrors);
+		},
+		(err: unknown) => {
+			stopFuzzing(err, options.expectedErrors);
+		}
 	);
 }
 
-export function addFuzzerOptionsForDryRun(
-	opts: string[],
-	shouldDoDryRun: boolean,
-	timeout = 5000
-): string[] {
-	const containsParameter = (params: string[], param: string): boolean => {
-		return params.some((curr) => curr.startsWith(param));
-	};
-	// Last occurrence of a parameter is used.
-	if (shouldDoDryRun) {
-		opts = opts.concat("-runs=0");
-	}
-	if (!containsParameter(opts, "-timeout")) {
-		const inSeconds = timeout / 1000;
-		opts = opts.concat(`-timeout=${inSeconds}`);
-	}
-	return opts;
-}
-
-export function startFuzzingNoInit(
-	fuzzFn: fuzzer.FuzzTarget,
-	fuzzerOptions: string[]
-) {
+export async function startFuzzingSync(options: Options) {
+	await initFuzzing(options);
+	const fuzzFn = await loadFuzzFunction(options);
+	const fuzzerOptions = addFuzzerOptionsForDryRun(
+		options.fuzzerOptions,
+		options.dryRun
+	);
 	Fuzzer.startFuzzing(fuzzFn, fuzzerOptions);
 }
 
 export async function startFuzzingAsync(options: Options) {
 	await initFuzzing(options);
-	const fuzzFn: fuzzer.FuzzTarget = await loadFuzzFunction(options);
+	const fuzzFn = await loadFuzzFunction(options);
 	return startFuzzingAsyncNoInit(
 		fuzzFn,
 		addFuzzerOptionsForDryRun(options.fuzzerOptions, options.dryRun)
```

Comment:
`startFuzzingSync` and this `Async` variant have a large overlap as they're already smallish functions, I'm uncertain whether we can extract some more into a common function to minimize code duplication. WDYT?

What we should do at least is keep them consistent and extract the `addFuzzerOptionsForDryRun` here as you did for the other.

Reply:
I also wondered about that but hesitated as they seemed to be subtly different. Regardless, I gave it a try and the result looks quite nice, see the additional commit. WDYT?

Reply:
I think that's a lot cleaner!